### PR TITLE
tech(encrypt): add pattern logic to encrypt the url

### LINKED ISSUE #18: encrypt: add pattern to the encrypt method
Add pattern logic to the encryption

### PR DESCRIPTION
**Pull request by shangan23**
implemented pattern logic to the encryption method

Request:
```js
encrypt('https://testapp.com/modules/provider/7d8bb2ff-04d8-4c3c-96c4-faafb6321432/20', {
    pattern: 'https://testapp.com/modules/provider/:id/:obj'
})
```

Output:
```
https://testapp.com/modules/provider/U2FsdGVkX19Tp1L2u3Suu8mMRJDdQKcfGx9iw4oI9aJHBXUcfp1L2u3Sg6BpNSKKH0XCYBzJp1L2u3Sp1L2u3Ses1L2a3S4hDieDcT5jn06oOQ1cp1L2u3SshZjAe1Q2u3A4le1Q2u3A4l/U2FsdGVkX18F790imrKGp1L2u3SD2KZwOTc85GVrZfACNQ3CMe1Q2u3A4l
```

Escaping special character from encryption:
https://stackoverflow.com/a/61077372/1929115

**Sample Code**

```js
// Original Text
var ciphertext = 'asda+dasd+asdas/asdasd/sadasdasd/dadasd=adsasda=dasd=';

// Replaced Text
var dataString = ciphertext.replace(/\+/g,'p1L2u3S').replace(/\//g,'s1L2a3S4h').replace(/=/g,'e1Q2u3A4l');
console.log(dataString);

// Back to Original Text
ciphertext = dataString.replace(/p1L2u3S/g, '+' ).replace(/s1L2a3S4h/g, '/').replace(/e1Q2u3A4l/g, '=');
console.log(ciphertext);
```

### CODE REVIEW

File: src/encrypt.js [hunk 1]
```diff
@@ -1,19 +1,30 @@
...
+ 
+  if(pattern === null)
+    throw new Error('pattern is missing')
 
-  return "https://testapp.com/modules/provider/WTq8zYcZfaWVvMncigHqwQ==";
+  // extract pattern from the url
+  const parsedPatternURL = Url(pattern);
+  const parsedGivenURL = Url(givenUrl);
+
+  const patternPathnameArray = parsedPatternURL.pathname.split("/");
```
Comment (ramsunvtech) [inline src/encrypt.js]: keep this as `pathPatternChunks`

File: src/encrypt.js [hunk 2]
```diff
@@ -1,19 +1,30 @@
...
+  const givenURLPathnameArray = parsedGivenURL.pathname.split("/");
```
Comment (ramsunvtech) [inline src/encrypt.js:16]: keep this as `givenPathChunks`

File: src/encrypt.js (hunk 1 above)
Comment (shangan23) [reply src/encrypt.js:15]: patternPathArrayChunk will do?

File: src/encrypt.js (hunk 2 above)
Comment (shangan23) [reply src/encrypt.js:16]: `givenPathArrayChunk` will do?

File: src/encrypt.js [hunk 3]
```diff
@@ -1,19 +1,30 @@
...
+
+  const filteredPatterns = patternPathnameArray.filter((param)=> param.indexOf(':') === 0);
+
+  filteredPatterns.map((p) => { 
+    let encryptedString = CryptoJS.AES.encrypt(givenURLPathnameArray[patternPathnameArray.indexOf(p)],secret).toString();
+    //to escape `/` (forward slash) from encryption
+    let escapedEncryptedString = encryptedString.replace(/\+/g,'p1L2u3S').replace(/\//g,'s1L2a3S4h').replace(/=/g,'e1Q2u3A4l');
+    givenURLPathnameArray[patternPathnameArray.indexOf(p)] = escapedEncryptedString
+  })
```
Comment (ramsunvtech) [inline src/encrypt.js]: rather doing string manipulation.

try some thing like below.

```
function encrypt(strings, ...encryptItems) {
    console.log('strings: ', strings)
    console.log('encryptItems: ', encryptItems)
    const iterationCount = Math.max(strings.length, encryptItems.length)
    let encryptedURl = ''
    const encoder = (s) => `#-${s}-#`
    for(let i=0;i<iterationCount;i++) {
        let encoded = encryptItems[i] ? encoder(encryptItems[i]) : ''
        encryptedURl = encryptedURl + (strings[i] || '') + encoded
    }

    console.log('encryptedURl: ', encryptedURl)
}

encrypt`https://testapp.com/modules/provider/${'7d8bb2ff-04d8-4c3c-96c4-faafb6321432/20'}/${'25'}`
```

node above-script.js
strings:  [ 'https://testapp.com/modules/provider/', '/', '' ]
encryptItems:  [ '7d8bb2ff-04d8-4c3c-96c4-faafb6321432/20', '25' ]
encryptedURl:  https://testapp.com/modules/provider/#-7d8bb2ff-04d8-4c3c-96c4-faafb6321432/20-#/#-25-#

Comment (ramsunvtech) [reply src/encrypt.js]: image

File: src/encrypt.js (hunk 1 above)
Comment (shangan23) [reply src/encrypt.js]: done

File: src/encrypt.js (hunk 3 above)
Comment (shangan23) [reply src/encrypt.js]: We need to discuss on this, please

**Review by ramsunvtech — APPROVED** [review]
Need to revisit later

**Pull request merged**